### PR TITLE
Pre-aggregation cache implementation for numeric monitoring plugins

### PR DESCRIPTION
**Pull request by akindyakov**
Pre-aggregation cache implementation for numeric monitoring plugins.

For the most of monitoring data some aggregations are going to be applied on the user side. To do analytics you rarely need to see all points on the graph. It means for such paths (unique keys for points subsets) particular points means not much. And to reduce a disk usage and a network traffic some pre-aggreagation could be applied on `osquery` side.

This PR is implementation of such pre-aggreagtion.

It based of PR #4626 , pay attention only to file `pre_aggregation_cache.h` and to file `pre_aggregation_cache.cpp`.

### CODE REVIEW

File: osquery/numeric_monitoring/pre_aggregation_cache.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+#include <unordered_map>
+
+#include <osquery/numeric_monitoring.h>
+
+namespace osquery {
+
+namespace monitoring {
+
+class Point {
```
Comment (theopolis) [inline osquery/numeric_monitoring/pre_aggregation_cache.h]: nitpick, simple docblocks for `Point`?

Comment (akindyakov) [reply osquery/numeric_monitoring/pre_aggregation_cache.h]: Sure, thank you!

File: osquery/numeric_monitoring/pre_aggregation_cache.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+bool Point::tryToAggregate(const Point& new_point) {
+  if (path_ != new_point.path_) {
+    LOG(WARNING) << "Pre-aggregation is no possible, point paths is not equal. "
+                    "That should not happen at all. Previous path is "
+                 << boost::io::quoted(path_) << ", new one is "
+                 << boost::io::quoted(new_point.path_);
+    return false;
+  }
+  if (pre_aggregation_type_ != new_point.pre_aggregation_type_) {
+    LOG(WARNING) << "Pre-aggregation is no possible, PreAggregationType type "
```
Comment (muffins) [inline osquery/numeric_monitoring/pre_aggregation_cache.cpp]: Maybe same thing here.

File: osquery/numeric_monitoring/pre_aggregation_cache.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    // nothing to do, the type is invalid
+    LOG(ERROR) << "Invalid Pre-aggregation type "
+               << boost::io::quoted(to<std::string>(pre_aggregation_type_));
+    return false;
+  }
+  return true;
+}
+
+void PreAggregationCache::addPoint(Point point) {
+  auto previousIndex = points_intex_.find(point.path_);
```
Comment (muffins) [inline osquery/numeric_monitoring/pre_aggregation_cache.cpp]: nit: I think `snake_case` for local variables?

File: osquery/numeric_monitoring/pre_aggregation_cache.h [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  void addPoint(Point point);
+
+  std::vector<Point> takePoints();
+
+  std::size_t size() const noexcept {
+    return points_.size();
+  }
+
+ private:
+  std::unordered_map<std::string, std::size_t> points_intex_;
```
Comment (muffins) [inline osquery/numeric_monitoring/pre_aggregation_cache.h]: Was this supposed to be `intex`? Seems like maybe it was intended to be `index`?

File: osquery/numeric_monitoring/tests/pre_aggregation_cache_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+#include <gtest/gtest.h>
+
+#include <osquery/tests/test_util.h>
+
+#include <osquery/numeric_monitoring/pre_aggregation_cache.h>
+
+namespace osquery {
+
+GTEST_TEST(PreAggregationPoint, tryToUpdate_same_path_none) {
```
Comment (muffins) [inline osquery/numeric_monitoring/tests/pre_aggregation_cache_tests.cpp]: Hm, I think we tend to use `TEST_F` for test functions as opposed to `GTEST_TEST`? Trying to track down why though.

File: osquery/numeric_monitoring/pre_aggregation_cache.cpp (hunk 3 above)
Comment (akindyakov) [reply osquery/numeric_monitoring/pre_aggregation_cache.cpp]: Yes, sure. Thank you!

File: osquery/numeric_monitoring/pre_aggregation_cache.h (hunk 4 above)
Comment (akindyakov) [reply osquery/numeric_monitoring/pre_aggregation_cache.h]: Yep.. mistype 🤕

File: osquery/numeric_monitoring/tests/pre_aggregation_cache_tests.cpp (hunk 5 above)
Comment (akindyakov) [reply osquery/numeric_monitoring/tests/pre_aggregation_cache_tests.cpp]: I saw both in the code, so I've chosen easiest one.

**Pull request merged**